automates/utils/method_dispatch.py

```diff
@@ -0,0 +1,12 @@
+from functools import singledispatch, update_wrapper
+
+# TODO Check if this can be removed now that we use python 3.8
+def methdispatch(func):
```

Comment:
@dpdicken Can you explain to me what is going on here? Simple `@singledispatch` wrappers have worked fine for me for some time, even with multiple args and kwargs. The only limitation seems to be that dispatch can only operate on the first argument. Are you doing something else here? The Python [docs](https://docs.python.org/3.8/library/functools.html#functools.singledispatch) seem to suggest that additional args and kwargs will work just fine.

Reply:
The issue is I wanted to use single dispatch on methods in a class in which their first argument is self! So, it breaks the normal use of single dispatch for class methods because you can't have variable types of self being passed in. So, this function allows you to use single dispatch on class methods by working around that problem. They actually implemented a version of this "methdispatch" in functors in python 3.8 to avoid this problem, which is why I have the todo there to check if that will work instead.

Reply:
Cool, yeah I just found the [docs](https://docs.python.org/3/library/functools.html#functools.singledispatchmethod) for `@singledispatchmethod` that was added in 3.8 as you noted in your comment (and your TODO). Can you make the switch to using that implementation for dispatch?